core/bin/qdrant/migrate_embedder.rs

```diff
@@ -0,0 +1,982 @@
+                        );
+
+                        Ok(ps)
+                    }
+                    None => Err(anyhow!("DataSource embedding error: Missin chunk")),
+                })
+                .collect::<Result<Vec<_>>>()?;
+
+            match qdrant_client
+                .upsert_points(
+                    &shadow_embedder_config,
+                    &ds.internal_id().to_string(),
+                    points_to_upsert,
+                )
+                .await
+            {
+                Ok(_) => (),
+                Err(e) => {
+                    if retry < 3 {
+                        retry += 1;
+                        utils::error(&format!(
+                            "Error migrating points (write): retry={} error={:?}",
+                            retry, e
+                        ));
+                        tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
+                        continue;
+                    } else {
+                        Err(e)?
+                    }
+                }
+            }
+        }
+
+        total += count;
+
+        if iterations % 8 == 0 {
+            utils::info(&format!(
+                "Migrated points: count={} total={} latency_ms={}",
+                count,
+                total,
+                utils::now() - now
+            ));
+        }
+
+        page_offset = scroll_results.next_page_offset;
+        if page_offset.is_none() {
+            break;
+        }
+        retry = 0;
+        iterations += 1;
+    }
+
+    utils::info(&format!("Done migrating: total={}", total));
+    Ok(())
+}
+
+async fn commit_shadow_embedder(
+    store: Box<dyn Store + Sync + Send>,
+    data_source_internal_id: String,
+) -> Result<()> {
+    let mut ds = match store
+        .load_data_source_by_internal_id(&data_source_internal_id)
+        .await?
+    {
+        Some(ds) => ds,
+        None => Err(anyhow!("Data source not found"))?,
+    };
+
+    let mut config = ds.config().clone();
+
+    // Switch embedder and shadow embedder.
+    config.embedder_config = match config.embedder_config.shadow_embedder {
+        Some(shadow_embedder) => EmbedderDataSourceConfig {
+            embedder: shadow_embedder.clone(),
+            shadow_embedder: Some(config.embedder_config.embedder),
+        },
+        None => Err(anyhow!("No shadow embedder to commit"))?,
+    };
+
+    ds.update_config(store, &config).await?;
+
+    utils::info(&format!(
+        "Updated data source: \
+            data_source_internal_id={}  data_source_id={} cluster={} embedder_config={} shadow_embedder_config={}",
+        ds.internal_id(),
+        ds.data_source_id(),
+        ds.main_qdrant_cluster().to_string(),
+        ds.embedder_config(),
+        match ds.shadow_embedder_config() {
+            Some(sec) => sec.to_string(),
+            None => "none".to_string(),
+        }
+    ));
+    Ok(())
+}
+
+async fn refresh_chunk_count_for_updated_documents(
```

Comment:
We can't access this one if we're doing a manual migration. Should we make it a command or add it at the end of `commit_shadow_embedder` instead of in `migrate` ?

Reply:
We can't really handle this in `commit_shadow_embedder` because we need the timestamp of when the shadow embedder was added to accurately determine the timeframe for refreshing the documents.

Reply:
I see :+1: